### PR TITLE
CLDC-3190 Add lettings BU for 24/25

### PR DESCRIPTION
**Pull request by kosiakkatrina**
Duplicate 2023 lettings bulk upload into 2024 lettings bulk upload and map questions to the new field numbers
Update duplicate log fields to sum individual charges and rent instead of using tcharge, because we no longer ask for tcharge in the template (field_132)

### CODE REVIEW

File: app/services/bulk_upload/lettings/year2024/csv_parser.rb [hunk 1]
```diff
@@ -0,0 +1,113 @@
+require "csv"
+
+class BulkUpload::Lettings::Year2024::CsvParser
+  include CollectionTimeHelper
+
+  FIELDS = 130
+  MAX_COLUMNS = 142
```
Comment (natdeanlewissoftwire) [inline app/services/bulk_upload/lettings/year2024/csv_parser.rb]: do we not want to reduce this by 4 as there are 4 fewer fields?

File: app/services/bulk_upload/lettings/year2024/csv_parser.rb [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+      0
+    end
+  end
+
+  def col_offset
+    with_headers? ? 1 : 0
+  end
+
+  def cols
+    @cols ||= ("A".."EL").to_a
```
Comment (natdeanlewissoftwire) [inline app/services/bulk_upload/lettings/year2024/csv_parser.rb]: similarly, if there are fewer fields we shouldn't go all the way to EL right?

File: app/services/bulk_upload/lettings/year2024/csv_parser.rb [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  def wrong_template_for_year?
+    collection_start_year_for_date(first_record_start_date) != FORM_YEAR
+  rescue Date::Error
+    false
+  end
+
+private
+
+  def default_field_numbers
+    (1..130).map { |h| h.present? && h.to_s.match?(/^[0-9]+$/) ? "field_#{h}" : "field_blank" }
```
Comment (natdeanlewissoftwire) [inline app/services/bulk_upload/lettings/year2024/csv_parser.rb]: use FIELDS here?

File: app/services/bulk_upload/lettings/year2024/row_parser.rb [hunk 4]
```diff
@@ -0,0 +1,1468 @@
+class BulkUpload::Lettings::Year2024::RowParser
+  include ActiveModel::Model
+  include ActiveModel::Attributes
+  include InterruptionScreenHelper
+
+  QUESTIONS = {
+    field_1: "Which organisation owns this property?",
```
Comment (natdeanlewissoftwire) [inline app/services/bulk_upload/lettings/year2024/row_parser.rb:7]: I haven't checked these, happy to trust you if you're happy they're right 😅

File: app/services/bulk_upload/lettings/year2024/row_parser.rb [hunk 5]
```diff
@@ -0,0 +1,1468 @@
...
+
+  validates :field_5,
+            presence: {
+              if: proc { supported_housing? },
+              message: I18n.t("validations.not_answered", question: "scheme code"),
+              category: :setup,
+            },
+            on: :after_log
+
+  validates :field_6,
```
Comment (natdeanlewissoftwire) [inline app/services/bulk_upload/lettings/year2024/row_parser.rb:332]: why are we making validation changes here? I thought all that would be left to separate tickets...

File: app/services/bulk_upload/lettings/year2024/row_parser.rb [hunk 6]
```diff
@@ -0,0 +1,1468 @@
...
+      errors.add(:field_122, error_message) unless general_needs? # household_charge
+      errors.add(:field_125, error_message) # brent
+      errors.add(:field_126, error_message) # scharge
+      errors.add(:field_127, error_message) # pscharge
+      errors.add(:field_128, error_message) # chcharge
+    end
+  end
+
+  def field_mapping_for_errors
+    {
```
Comment (natdeanlewissoftwire) [inline app/services/bulk_upload/lettings/year2024/row_parser.rb:851]: why have we removed tcharge and offered from this?

**Review by natdeanlewissoftwire — APPROVED** [review]
looks good! Just a couple things I wasn't sure about but I'm sure they're intentional

File: app/services/bulk_upload/lettings/year2024/csv_parser.rb (hunk 2 above)
Comment (kosiakkatrina) [reply app/services/bulk_upload/lettings/year2024/csv_parser.rb]: You're right! It looks like it only goes up to EA!

File: app/services/bulk_upload/lettings/year2024/csv_parser.rb (hunk 1 above)
Comment (kosiakkatrina) [reply app/services/bulk_upload/lettings/year2024/csv_parser.rb]: So EL is 142, EA is 131 so will update it to that. Thank you!

File: app/services/bulk_upload/lettings/year2024/row_parser.rb (hunk 4 above)
Comment (kosiakkatrina) [reply app/services/bulk_upload/lettings/year2024/row_parser.rb:7]: Yeah, there were a lot of mappings, I hope they're right and I don't expect you to check every single one of them 😅

File: app/services/bulk_upload/lettings/year2024/row_parser.rb (hunk 5 above)
Comment (kosiakkatrina) [reply app/services/bulk_upload/lettings/year2024/row_parser.rb:332]: I think if you're referring to the location fields, it's cause we now only have use the new scheme and location fields, so we don't need the validation which determines which scheme/location/management group fields get answered, but we still want to validate the location. So this is just the simplified version of that. If you have other validations in mind that shouldn't be a part of this let me know!

File: app/services/bulk_upload/lettings/year2024/row_parser.rb (hunk 6 above)
Comment (kosiakkatrina) [reply app/services/bulk_upload/lettings/year2024/row_parser.rb:851]: We're no longer asking offered question and tcharge I'll add back and will map to the other charges fields (we're no longer ask tcharge either)

**Pull request merged**